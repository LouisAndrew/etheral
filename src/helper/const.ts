import { Product } from './schema/product';
import { Order } from './schema/order';

const mockProduct1: Product = {
    pid: '1234',
    name: 'Mock product',
    slug: 'ctg1/product1',
    category: 'ctg1',
    idrPrice: 23000,
    collection: 'fall',
    productImages: [],
    urls: [],
    weight: 20,
    amount: 12,
};

export const mockProducts: Product[] = [mockProduct1];

export const mockOrder: Order = {
    oid: 'asdadwda',
    buyerName: 'Jane Doe',
    buyerEmail: 'jane@doe',
    buyerPhone: 123412412,
    buyerUId: 'admin',
    buyerAddr: 'Storkowerstrasse',
    buyerPostal: 1234143,
    total: 32000,
    currency: 'IDR',
    date: new Date(),
    via: 'va',
    channel: 'bni',
    products: [{ pid: mockProduct1.pid, amount: 12 }],
    paid: false,
    delivered: false,
    transactionData: {
        sessionId: '1234',
        paymentNo: '12',
        paymentName: 'Jane Doe - BNI',
        expired: '1231241412',
        fee: 3000,
    },
    shippingMethod: 'OKE',
};

export const mockOrders: Order[] = [mockOrder];

export const mockOrderShipped: Order = {
    ...mockOrder,
    delivered: true,
    shippingData: {
        shippedDate: new Date(),
        trackingCode: '12342',
        shippedBy: 'admin1',
    },
};
