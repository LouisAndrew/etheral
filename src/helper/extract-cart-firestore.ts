import { intersectionBy } from 'lodash';

import { SetCartArgs } from 'state/actions/cart';
import { InCart } from './schema/firebase-user';
import { Product } from './schema/product';

export default ({
    firestoreCartData,
    allProducts,
}: {
    firestoreCartData: InCart;
    allProducts: Product[];
}): SetCartArgs => {
    const inCartProduct = intersectionBy(allProducts, firestoreCartData, 'pid');
    return inCartProduct.map(product => ({
        product,
        amount: intersectionBy(firestoreCartData, [product], 'pid')[0].amount, // get the intersection between the product -> pid is unique, so the function is guaranteed to return one item.
    }));
};
