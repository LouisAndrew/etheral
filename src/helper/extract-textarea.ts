import { snakeCase } from 'lodash';

/**
 * function to extract options from a textarea
 * @param textarea textarea input (e.g sizes: S, M, L)
 */
export const extractTextArea = (textarea: string) =>
    textarea.split(',').map(text => snakeCase(text).toUpperCase());
