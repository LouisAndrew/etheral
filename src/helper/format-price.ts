export const formatPrice = (price: number | string) => {
    const value =
        typeof price === 'string'
            ? parseFloat(price.replace('IDR ', ''))
            : price;
    return new Intl.NumberFormat('id').format(value);
};
