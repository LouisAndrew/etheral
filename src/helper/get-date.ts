const pad2 = (n: number): string => (n < 10 ? `0${n}` : n.toString());

const getDate = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7);
    const minutes = pad2(date.getUTCMinutes());

    return year + '-' + month + '-' + day + ' ' + hours + ':' + minutes;
};

const getDateIpaymu = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = pad2(date.getUTCMonth() + 1); // getMonth() is zero-indexed, so we'll increment to get the correct month number
    const day = pad2(date.getUTCDate());
    const hours = pad2(date.getUTCHours() + 7); // UTC offset for WIB => +7
    const minutes = pad2(date.getUTCMinutes());
    const seconds = pad2(date.getUTCSeconds());

    return `${year}${month}${day}${hours}${minutes}${seconds}`;
};

const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const getDateReadable = (date: Date) => {
    const year = date.getUTCFullYear();
    const month = date.getUTCMonth();
    const day = date.getUTCDate();

    return `${day} ${monthNames[month]} ${year}`;
};

export { getDate, getDateIpaymu, getDateReadable };
