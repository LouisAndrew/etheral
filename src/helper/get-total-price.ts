import { Product } from 'helper/schema';
import { withDiscount } from './with-discount';
import { Currencies } from 'state/reducers/currency-reducer';

type HelperArgs = {
    [key: string]: any;
    amount: number;
    product: Product;
};

const getTotalPriceIdr = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, idrPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return acc + amount * price;
    }, 0);
};

const getTotalPriceAud = (args: HelperArgs[]) => {
    return args.reduce((acc, curr) => {
        const { amount, product } = curr;

        // destructure prices prop
        const {
            prices: { discountPercentage, ausPrice },
        } = product;

        // get price.
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return acc + amount * price;
    }, 0);
};

/**
 * function to get price (formatted ?) based on currency
 * @param args product name and prices + its discount percentage.
 * @param currency current currency state
 */
const getPrice = (args: HelperArgs, currency: Currencies) => {
    const {
        product: { prices },
        amount,
    } = args;

    const { discountPercentage, idrPrice, ausPrice } = prices;

    // handle IDR.
    if (currency === Currencies.IDR) {
        const price =
            discountPercentage > 0
                ? withDiscount(idrPrice, discountPercentage)
                : idrPrice;

        return amount * price;
    } else {
        const price =
            discountPercentage > 0
                ? withDiscount(ausPrice, discountPercentage)
                : ausPrice;

        return amount * price;
    }
};

export { getTotalPriceIdr, getTotalPriceAud, getPrice };
