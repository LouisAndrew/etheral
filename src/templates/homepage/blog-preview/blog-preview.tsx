import React from 'react';
import Img, { FluidObject } from 'gatsby-image';
import { Link } from '@reach/router';

import { Box, Flex, Heading, Text, Button } from 'rebass';

import { Blog } from 'helper/schema/blog';
import { getDateReadable } from 'helper/get-date';

type Props = {
    blog: Blog;
};

const BlogPreview: React.FC<Props> = ({ blog }) => {
    const { title, summary, date, image, slug } = blog;

    return (
        <Box px={[6, 6, 8]} mb={[4, 4, 5, 6]}>
            <Flex
                flexDirection={['column', 'column', 'column', 'row']}
                alignItems={[
                    'unset',
                    'unset',
                    'center',
                    'flex-start',
                    'center',
                ]}
                justifyContent="space-between"
            >
                {/* render preview image. */}
                <Box width={['100%', '100%', '80%', '48%', '48%']}>
                    {image && (
                        <Img
                            fluid={image.childImageSharp.fluid as FluidObject}
                        />
                    )}
                </Box>
                {/* actual preview. */}
                <Box width={['100%', '100%', '80%', '48%', '48%']}>
                    {/* Render date here. */}
                    <Text as="h5" variant="h5" mt={[4, 4, 6, 0]}>
                        {getDateReadable(date)}
                    </Text>
                    {/* title */}
                    <Heading as="h3" variant="h3" data-testid="title" mb={[4]}>
                        {title}
                    </Heading>
                    {/* summary */}
                    <Text as="article" variant="body" data-testid="summary">
                        {summary}
                    </Text>
                    <Link to={`/blogs/${slug}`}>
                        <Button
                            width={['100%', '100%', 'fit-content']}
                            mt={[4]}
                            mb={[7]}
                        >
                            Read More
                        </Button>
                    </Link>
                </Box>
            </Flex>
        </Box>
    );
};

export { BlogPreview };
