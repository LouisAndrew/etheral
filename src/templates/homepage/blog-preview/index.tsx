import React from 'react';

import { useAllBlogs } from 'helper/use-all-blogs';
import { BlogPreview as BlogPreviewEl } from './blog-preview';

const BlogPreview: React.FC<{}> = () => {
    const { sort } = useAllBlogs();

    const sorted = sort();

    // render preview if there's at least one blog post.
    if (sorted.length > 0) {
        const newest = sorted[sorted.length - 1];
        return <BlogPreviewEl blog={newest} />;
    } else {
        return <></>;
    }
};

export default BlogPreview;
