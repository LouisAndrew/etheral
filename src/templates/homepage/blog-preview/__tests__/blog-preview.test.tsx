import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { BlogPreview } from '../blog-preview';
import { mockBlog } from 'helper/const';

describe('BlogPreview', () => {
    const Element = <BlogPreview blog={mockBlog} />;

    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('should render preview with summary and title correctly', () => {
        const { queryByTestId } = render(Element);

        const TitleElement = queryByTestId('title');
        const SummaryElement = queryByTestId('summary');

        if (TitleElement && SummaryElement) {
            expect(TitleElement).toHaveTextContent(mockBlog.title);
            expect(SummaryElement).toHaveTextContent(mockBlog.summary);
        } else {
            fail();
        }
    });

    it('matches snapshot', () => {
        const run = true;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
