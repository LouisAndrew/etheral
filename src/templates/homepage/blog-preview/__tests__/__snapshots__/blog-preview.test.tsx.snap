// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPreview matches snapshot 1`] = `
<div
  className="css-702kq"
>
  <div
    className="css-1u841xy"
  >
    <div
      className="css-1x87397"
    />
    <div
      className="css-1x87397"
    >
      <h5
        className="css-4fmxwb"
      >
        10 September 2020
      </h5>
      <h3
        className="css-d2e41n"
        data-testid="title"
      >
        Mock title
      </h3>
      <article
        className="css-vurnku"
        data-testid="summary"
      >
        Mock summary
      </article>
      <a
        href="/blogs/sample-blog-post"
        onClick={[Function]}
      >
        <button
          className="css-1iwxm44"
        >
          Read More
        </button>
      </a>
    </div>
  </div>
</div>
`;
